scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua

```diff
@@ -3,117 +3,534 @@
-    local sLuna         = trade:hasItemQty(xi.item.SEEDSPALL_LUNA, 1)
-    local sAstrum       = trade:hasItemQty(xi.item.SEEDSPALL_ASTRUM, 1)
-    local acpMission    = player:getCurrentMission(xi.mission.log_id.ACP)
-    local hasCrimsonKey = player:hasKeyItem(xi.ki.CRIMSON_KEY)
-    local lastCrimson   = player:getCharVar('LastCrimsonKey') -- When last Crimson key was obtained
+    local now                   = os.time()
+    local nextMidnight          = getMidnight()
+    local count                 = trade:getItemCount()
+    local finishedACP           = player:getCurrentMission(xi.mission.log_id.ACP) == xi.mission.id.acp.A_CRYSTALLINE_PROPHECY_FIN
+    local finishedAMK           = player:getCurrentMission(xi.mission.log_id.AMK) == xi.mission.id.amk.A_MOOGLE_KUPO_DETAT_FIN
+    local finishedASA           = player:getCurrentMission(xi.mission.log_id.ASA) == xi.mission.id.asa.A_SHANTOTTO_ASCENSION_FIN
+    local hasMoogleKey          = player:hasKeyItem(xi.ki.MOOGLE_KEY)
+    local kit                   = player:getCharVar('ASA_kit')
+    local hasMoogleItems        = kit ~= 0 and
+                                    trade:hasItemQty(kit, 1) and
+                                    count == 1
+    local hasCrimsonItems       = trade:hasItemQty(xi.item.SEEDSPALL_LUX, 1)        and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_LUNA, 1)     and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_ASTRUM, 1)   and
+                                    count == 3
+    local hasWhiteCoralItems    = trade:hasItemQty(xi.item.ORCISH_PLATE_ARMOR, 1)   and
+                                    trade:hasItemQty(xi.item.QUADAV_BACKSCALE, 1)   and
+                                    trade:hasItemQty(xi.item.YAGUDO_CAULK, 1)       and
+                                    count == 3
 
+    -- Crimson Key: Trade Seedspall's Lux, Luna, Astrum
     if
-        xi.settings.main.ENABLE_ACP == 0 and
-        xi.settings.main.ENABLE_AMK == 0 and
-        xi.settings.main.ENABLE_ASA == 0
+        hasCrimsonItems and
+        finishedACP
     then
-        player:showText(npc, ID.text.GET_LOST)
-
-    -- Crimson Key: Trade Seedspall's Lux, Luna, Astrum
-    else
+        local ki = xi.ki.CRIMSON_KEY
         if
-            xi.settings.main.ENABLE_ACP == 1 and
-            sLux and
-            sLuna and
-            sAstrum and
-            count == 3 and
-            acpMission >= xi.mission.id.acp.GATHERER_OF_LIGHT_I and
-            not hasCrimsonKey and
-            os.time() > lastCrimson
-        then -- and timer stuff here) then
+            not player:hasKeyItem(ki) and
+            now >= player:getCharVar('LastCrimsonKey')
+        then
             player:tradeComplete()
-            player:addKeyItem(xi.ki.CRIMSON_KEY)
-            player:setCharVar('LastCrimsonKey', getMidnight())
+            player:addKeyItem(ki)
+            player:setCharVar('LastCrimsonKey', nextMidnight)
```

Comment:
With the different key trades, you can probably write some helper funcs to deal with the repeated logic

Reply:
Fair point